### DISCUSSION TITLE
Offline Transaction Signing

### DISCUSSION THREAD

**Discussion by Ealrann** [Chia-Network/chia-blockchain#7784] [Ideas]
Hello Chia team :hand:,

Do you have any plans about offline signing?

I mean by that the ability to:
1. Create and sign a transaction from a cold wallet (offline computer with the private key)
2. Publish this transaction on the network with an online full node.

I don't find information about this, but I think it's an important feature since it would allows us to safely send money from a cold wallet. Even more, enthusiast users could easily create their own hardware wallet.

**Comment by SlugFiller** [comment]
AFAIK, there are plans for supporting hardware wallets, with no need for "enthusiast". There have been talks going on with Ledger to add support for Chia, and vice versa.

Currently, Chia's main issue in terms of hardware wallets is the fact that it uses EIP-2334 for key derivation instead of BIP32. EIP-2334 is part of Eth2, the planned evolution of Etherium, and hardware wallets generally do not have it implemented. Trezor, specifically, have previously noted they're holding on implementation of Eth2 support until someone makes a verified reference implementation.

**Reply by Ealrann** [reply]
Thanks for this info, that's cool that Chia use a modern key derivation. 

But my question is more about offline signing with the official tool, any news about that ?

**Reply by SlugFiller** [reply]
From what I can tell, it's not possible directly through the client. From what I can tell, there are no calls to `Wallet._generate_unsigned_transaction` outside of `Wallet.generate_signed_transaction`. It might be possible via the Python API, but that's a tad advanced. Even the RPC interface only allows creating a signed transaction, and not an unsigned one, although it at least allows separating transaction signing from broadcast.

From what I gather, there's also some technical issue in making a full-on cold wallet. A key requirement is that you have an active wallet that can at least read the status of the wallet, as in, how much XCH is available (and accessible via the seed), and in which addresses. For the cold wallet to truly be cold, this information has to be gathered without having any access to the seed. But it seems like the key derivation algorithm makes it so that you can't even figure out the possible addresses in the wallet, without having partial access to the seed. And yet, at the same time, it's impossible to know which addresses are in use without having network access. So, figuring out all the addresses currently in use without allowing a machine with the seed to have network access, would require quite a bit of back-and-forth, with the cold wallet generating possible addresses, and the connected wallet checking which addresses are in use, and requesting more potential addresses.

This sort of back-and-forth is, of course, possible with a hardware wallet, but a lot more problematic with a human running back and forth with a disk-on-key.

**Comment by mariano54** [comment]
If you know how to code, you can use this:
https://github.com/Chia-Network/offline-signing-demo
